src/main/scala/au/com/agiledigital/toolform/model/Project.scala

```diff
@@ -0,0 +1,114 @@
+                         volumes: Option[Seq[Volume]],
+                         componentGroups: Option[Seq[ComponentGroup]])
+
+/**
+  * The network topology of the environment explains how elements will be related/connected.
+  * @param links the links between components, other components and resources.
+  * @param edges the edges that will make the components and resources available outside the project.
+  */
+final case class Topology(links: Seq[Link], edges: Map[String, Edge])
+
+/**
+  * Points to an element of the project that is defined in another part of the project configuration.
+  * See Link, Volume, Location.
+  * @param refType type of the referenced object eg. "components", "resources"
+  * @param refId
+  * .  eg. "components.componentA"
+  */
+final case class Reference(refType: ReferenceType, refId: String) {
+  def resolve(project: Project): Option[ProjectElement] = refType match {
+    case ReferenceType.components => project.components.get(refId)
+    case ReferenceType.resources  => project.resources.get(refId)
+    case _                        => None
+  }
+  override def toString: String = s"$refType.$refId"
+}
+
+object Reference {
+  implicit val converter: ConfigConvert[Reference] =
+    ConfigConvert.viaStringOpt(
+      s => {
+        val referenceParts = s.split('.')
+        if (referenceParts.length == 2) {
+          val refType = referenceParts(0)
+          val refId   = referenceParts(1)
+          Some(Reference(ReferenceType.withName(refType), refId))
+        } else
+          throw new IllegalArgumentException(s"Failed to parse reference [$s] - it was not of format [type.id]")
+      },
+      _.toString
+    )
+}
+
+/**
+  * Kinds of references that may be used in project config.
+  */
+object ReferenceType extends Enumeration {
+  type ReferenceType = Value
+  val components, resources = Value
+}
+
+/**
+  * A link between two project elements.
+  *
+  * @param from the ProjectElement that will be linked into the to ProjectElement
+  * @param to   the ProjectElement that will be supplied the from ProjectElement
+  */
+final case class ResolvedLink(from: ProjectElement, to: ProjectElement)
+
+/**
+  * A link between two endpoints defined in the project.
+  * An endpoint may be a resource, a component, or another addressable location.
+  * E.g.
+  * {
+  *   from: resources.app_database
+  *   to: components.public_api
+  * }
+  * @param from The source endpoint that will be linked from.
+  * @param to The target endpoint that will be linked to.
+  */
+final case class Link(from: Reference, to: Reference) {
+  def resolve(project: Project): ResolvedLink = {
+    def invalidPathError(ref: Reference): Exception =
+      new IllegalArgumentException(s"""Could not resolve link to path [$ref]. Available components [${project.components.keys.mkString(",")}]""")
+
+    val maybeFrom = from.resolve(project)
+    val maybeTo   = to.resolve(project)
+    maybeFrom match {
```

Comment:
This can be written as:
```
(maybeFrom, maybeTo) match {
  case (None, None) =>
  case (None, _) =>
  case (_, None) =>
  case (Some(...), Some(...)) =>
}
```
which is probably cleaner.

That said, when you have the possibility of accumulating multiple errors, scalaz has a `ValidationNEL` that is designed to do just that. When accumulating, I'd recommend using `ValidationNEL[String, X]` instead.

Reply:
Would like this one change (to matching on the tuple) as part of the PR.